### PR TITLE
feat: Add type definitions for npm module `ospec`

### PR DESCRIPTION
**Pull request by maranomynet**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/ospec/ospec-tests.ts [hunk 1]
```diff
@@ -0,0 +1,208 @@
+// import o = require('ospec');
+import o, { Definer } from 'ospec'; // NOTE: this only works with  "esModuleInterop": true
+
+const exampleTypeUse1: o.Definer = () => {};
+const exampleTtypeUse2: Definer = () => {}; // NOTE: this only works with  "esModuleInterop": true
```
Comment (isiahmeadows) [inline types/ospec/ospec-tests.ts]:
```suggestion
const exampleTypeUse2: Definer = () => {}; // NOTE: this only works with  "esModuleInterop": true
```

File: types/ospec/ospec-tests.ts [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+        // $ExpectError
+        o(bool).notThrows('baz');
+
+        // `expected` must only be string or "Object constructor"
+        // $ExpectError
+        o(fn).throws(1);
+        // $ExpectError
+        o(fn).throws(1);
+
+        /*
```
Comment (isiahmeadows) [inline types/ospec/ospec-tests.ts]: See above - my solution should allow you to uncomment this section.

**Review by dead-claudia — CHANGES_REQUESTED** [review]
See the inline comments. Also, don't forget to test tests returning promises!

Note to DT maintainers: I'm the current maintainer of this library (it's part of the Mithril ecosystem and is used in core development).

File: types/ospec/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,106 @@
+// Type definitions for ospec 4.0
+// Project: https://github.com/MithrilJS/mithril.js/tree/next/ospec
+// Definitions by: Már Örlygsson <https://github.com/maranomynet>
+//                 Mike Linkovich <https://github.com/spacejack>
+//                 Isiah Meadows <https://github.com/isiahmeadows>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 3.2
+
+declare namespace Ospec {
```
[the remark this answers is not in the document]
Comment (maranomynet) [reply types/ospec/index.d.ts]: OK, I'll change that, but AFAIU this definition file doesn't export any global definitions ... or does it? ...or maybe it should?

(I find TypeScript's documentation quite confusing at times.)

File: types/ospec/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    type AssertionDescriber = (description: string) => void;
+
+    interface Spy<Args extends any[] = any[], Returns = undefined> {
```
[the remark this answers is not in the document]
Comment (maranomynet) [reply types/ospec/index.d.ts]: Good suggestion, and brings it closer to @spacejack 's original notation. This was me trying too hard to simplify things, but inappropriately.

File: types/ospec/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        args: Args;
+        /** List of arguments that were passed to the function each tine it was called */
+        calls: Args[];
+    }
+
+    interface BasicAssertions<T = any> {
+        /** Asserts that two values are strictly equal */
+        equals(expected: T): AssertionDescriber;
+        /** Asserts that two values are **not** strictly equal */
+        notEquals(value: any): AssertionDescriber;
```
[the remark this answers is not in the document]
Comment (maranomynet) [reply types/ospec/index.d.ts]: I think users may have valid reasons to test against seemingly obvious un-equalities.
Who are we to judge? :-)

If you feel strongly about this, I'll make this change.

File: types/ospec/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        (...args: Args): Returns;
+        /** The number of times the function has been called */
+        callCount: number;
+        /** The arguments that were passed to the function in the last time it was called */
+        args: Args;
+        /** List of arguments that were passed to the function each tine it was called */
+        calls: Args[];
+    }
+
+    interface BasicAssertions<T = any> {
```
[the remark this answers is not in the document]
Comment (maranomynet) [reply types/ospec/index.d.ts]: Interesting approach! I'd never have tought of that.

File: types/ospec/index.d.ts [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        // which errors at runtime.
+
+        /** Asserts that the function throws an error of a given type */
+        throws(error: string | NewableFunction): AssertionDescriber;
+
+        /** Asserts that the function does **not** throw an error of given type */
+        notThrows(error: string | NewableFunction): AssertionDescriber; // See above
+    }
+
+    type Definer = (
```
[the remark this answers is not in the document]
Comment (maranomynet) [reply types/ospec/index.d.ts]: Regarding the `{}`: Judging by the ospec source `done` accepts any non-nully, non-Errory value and cast it to string (`String(err)` in [ospec.js#L152](https://github.com/MithrilJS/mithril.js/blob/next/ospec/ospec.js#L152))

Should we not reflect that?
...or at least accept `string` also?

File: types/ospec/index.d.ts [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+    interface ObjectAssertions<T = any> extends BasicAssertions<T> {
+        /** Asserts that two objects are recursively equal */
+        deepEquals(expected: T): AssertionDescriber;
+        /** Asserts that two objects are **not** recursively equal */
+        notDeepEquals(value: object): AssertionDescriber;
+    }
+
+    interface FunctionAssertions<T = any> extends ObjectAssertions<T> {
+        // FIXME: Find a more appropriate type annotation for the `error` parameter
```
[the remark this answers is not in the document]
Comment (maranomynet) [reply types/ospec/index.d.ts]: TIL
:-)

File: types/ospec/index.d.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (maranomynet) [reply types/ospec/index.d.ts]: Also, I can't get your suggested signatures to work as intended.

Tsc starts complaining about normal usage like

```js
o('test', () => {});
// Type 'void' is not assignable to type 'PromiseLike<any>'
```

and if I allow the first signature return to `PromiseLike<any> | void` then this error appears:

```js
o('test', (done, timeout) => {});
// Parameter 'done' implicitly has an 'any' type.
// Parameter 'timeout' implicitly has an 'any' type.
```

For now, I'm thinking about going for this less-clever improvement:

```
type Definer = (
    done: (error?: Error | string | null) => void,
    timeout: (delay: number) => void,
) => void | PromiseLike<any>;
```

**Comment by maranomynet** [comment]
BTW, I don't think the failed Travis build is my fault, and affects most other PRs also.

From the log:
```
Error: In package.json: Dependency @vue/test-utils not in whitelist.
```

File: types/ospec/index.d.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (isiahmeadows) [reply types/ospec/index.d.ts]: Okay, let's just go with the less clever part.

> Regarding the {}: Judging by the ospec source done accepts any non-nully, non-Errory value and cast it to string (String(err) in https://github.com/MithrilJS/mithril.js/blob/next/ospec/ospec.js#L152 )
>
> Should we not reflect that?
> ...or at least accept string also?

It's more idiomatic to require actual errors in TS APIs and declaration files, even if the underlying library supports more generic stuff. It also follows the fact TS won't let you throw anything other than an `Error` instance, despite JS letting you throw just about anything.

File: types/ospec/index.d.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (isiahmeadows) [reply types/ospec/index.d.ts]: It *should*. TS *idiomatically* prefers upper-cased namespaces, but the grammar doesn't *mandate* it.

File: types/ospec/index.d.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (isiahmeadows) [reply types/ospec/index.d.ts]: Go ahead and make the change. Static types exist for a reason! 😉

File: types/ospec/tsconfig.json [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+        "strictNullChecks": true,
+        "baseUrl": "../",
+        "typeRoots": [
+            "../"
+        ],
+        "types": [],
+        "noEmit": true,
+        "forceConsistentCasingInFileNames": true,
+
+        "esModuleInterop": true
```
Comment (sheetalkamat) [inline types/ospec/tsconfig.json]: Remove this and fix the test.. This flag is not needed just for writing test

**Review by sheetalkamat — CHANGES_REQUESTED** [review]

**Pull request merged**